remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,562 @@
+  for (auto &block : func) {
+    // The machines rose from the ashes of the nuclear fire....
+    if (block.getTerminator()) {
+      curr_wl.push_back(&block);
+      // add block to map with initial entry set
+      block_map.insert({&block, LiveSet()});
+    }
+  }
+}
+
+// Visit the basic blocks in the worklist and update the block_map.
+void LiveSetBlockVisitor::Visit(void) {
+  bool progress = true;
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty() && progress) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    progress = false;
+    for (auto block : curr_wl) {
+      if (VisitBlock(block)) {
+        progress = true;
+        // Updates have been made; add predecessors to next rotation
+        for (llvm::BasicBlock *pred : predecessors(block)) {
+          next_wl.push_back(pred);
+        }
+      }
+      //LOG(INFO) << "LSBV: " << live->count() << " live slots";
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
+  }
+  LOG(INFO) << "Found " << to_remove.size() << " dead stores to remove";
+  RemoveDeadStores(to_remove);
+}
+
+bool LiveSetBlockVisitor::VisitBlock(llvm::BasicBlock *B) {
+  //block_map[B] = *init;
+  LiveSet live;
+  for (auto inst_it = B->rbegin(); inst_it != B->rend(); ++inst_it) {
+    auto inst = &*inst_it;
+    if (llvm::isa<llvm::ReturnInst>(inst) || llvm::isa<llvm::UnreachableInst>(inst)) {
+      live.set();
+    } else if (llvm::isa<llvm::BranchInst>(inst)) {
+      // update live from successors
+      for (llvm::BasicBlock *succ : successors(B)) {
+        // all successors must be default have been visited first
+        // b = U { s | for s in successors(b)}
+        live |= block_map[succ];
+      }
+    } else if (auto call_inst = llvm::dyn_cast<llvm::CallInst>(inst)) {
+      if (call_inst->getFunctionType() == lft) {
+        // mark as all live if called function has the same type as lifted function
+        // we're not able to see inside other paths, so we can't predict whether or
+        // not the callee might not use any of our slots
+        live.set();
+      }
+    } else if (llvm::isa<llvm::InvokeInst>(inst)) {
+      if (call_inst->getFunctionType() == lft) {
+        // mark as all live if invoked function has the same type as lifted function
+        // we're not able to see inside other paths, so we can't predict whether or
+        // not the callee might not use any of our slots
+        live.set();
+      }
+    } else if (llvm::isa<llvm::StoreInst>(inst)) {
+      if (auto scope = GetScopeFromInst(*inst)) {
+        if (!live.test(scope_to_offset.at(scope))) {
+          // Case 1: slot is marked dead by a previous store and we hit a store.
+          // (i.e. testing the slot at this offset for liveness returns false)
+          // This is a dead store, so we can add it to the chopping block.
+          to_remove.insert(inst);
```

Comment:
This is too soon. We can't mark anything for removal until the algorithm converges. Perhaps you can have a boolean parameter that, when `true`, adds stuff to the `to_remove` set. That way, we can run the data flow algorithm with `false` until it converges, then re-run on each block with `true` to collect the final set of instructions to remove.

Reply:
Also, you need to get at the size of the value being stored, and compare it to the size of the slot.